containers/src/Data/Map/Internal.hs

```diff
@@ -3535,8 +3536,30 @@ argSet (Bin sz kx x l r) = Set.Bin sz (Arg kx x) (argSet l) (argSet r)
 -- > fromSet undefined Data.Set.empty == empty
 
 fromSet :: (k -> a) -> Set.Set k -> Map k a
-fromSet _ Set.Tip = Tip
-fromSet f (Set.Bin sz x l r) = Bin sz x (f x) (fromSet f l) (fromSet f r)
+#ifdef __GLASGOW_HASKELL__
+fromSet f = runIdentity . fromSetA (coerce f)
+#else
+fromSet f = runIdentity . fromSetA (pure . f)
+#endif
+
+-- | \(O(n)\). Build a map from a set of keys and a function which for each key
+-- computes its value in an 'Applicative' context.
+--
+-- > fromSetA (\k -> pure $ replicate k 'a') (Data.Set.fromList [3, 5]) == pure (fromList [(5,"aaaaa"), (3,"aaa")])
+-- > fromSetA undefined Data.Set.empty == pure empty
+
+fromSetA :: Applicative f => (k -> f a) -> Set.Set k -> f (Map k a)
+fromSetA _ Set.Tip = pure Tip
+fromSetA f (Set.Bin sz x l r) =
+  flip (Bin sz x)
+    <$> fromSetA f l
+    <*> f x
```

Comment:
I would prefer to use `liftA2`, since it may be defined to be more efficient for the Applicative used.